db/view/view.cc

```diff
@@ -1719,37 +1720,62 @@ bool should_generate_view_updates_on_this_shard(const schema_ptr& base, const lo
 //
 // If the keyspace's replication strategy is a NetworkTopologyStrategy,
 // we pair only nodes in the same datacenter.
-// If one of the base replicas also happens to be a view replica, it is
-// paired with itself (with the other nodes paired by order in the list
+//
+// When use_legacy_self_pairing is enabled, if one of the base replicas
+// also happens to be a view replica, it is paired with itself
+// (with the other nodes paired by order in the list
 // after taking this node out).
 //
+// If the table uses tablets and the replication strategy is NetworkTopologyStrategy
+// and the replication factor in the node's datacenter is a multiple of the number
+// of racks in the datacenter, then pairing is rack-aware.  In this case,
+// all racks have the same number of replicas, and those are never migrated
+// outside their racks. Therefore, the base replicas are naturally paired with the
+// view replicas that are in the same rack, based on the ordinal position.
+// Note that typically, there is a single replica per rack and pairing is trivial.
+//
 // If the assumption that the given base token belongs to this replica
 // does not hold, we return an empty optional.
-static std::optional<gms::inet_address>
+std::optional<gms::inet_address>
 get_view_natural_endpoint(
+        locator::host_id me,
         const locator::effective_replication_map_ptr& base_erm,
         const locator::effective_replication_map_ptr& view_erm,
-        bool network_topology,
+        const locator::abstract_replication_strategy& replication_strategy,
         const dht::token& base_token,
         const dht::token& view_token,
         bool use_legacy_self_pairing,
+        bool use_tablets_basic_rack_aware_view_pairing,
         replica::cf_stats& cf_stats) {
     auto& topology = base_erm->get_token_metadata_ptr()->get_topology();
-    auto me = topology.my_host_id();
-    auto my_datacenter = topology.get_datacenter();
+    auto& my_location = topology.get_location(me);
+    auto& my_datacenter = my_location.dc;
+    auto* network_topology = dynamic_cast<const locator::network_topology_strategy*>(&replication_strategy);
+    auto rack_aware_pairing = use_tablets_basic_rack_aware_view_pairing && network_topology &&
+            (network_topology->get_replication_factor(my_datacenter) % topology.get_datacenter_racks().at(my_datacenter).size()) == 0;
     std::vector<locator::host_id> base_endpoints, view_endpoints;
```

Comment:
isn't it a problem that rack aware pairing can be suddenly disabled after RF increase?

Reply:
@bhalevy @nyh 

assume:

nodes = rack1(n11, n12), rack2(n21, n22), rack3(n31, n32)

RF=4, racks=3

replica lists sorted by rack:

base { n11, n21, n31, n32 }
view { n12, n21, n22, n31 }

base has a higher load on rack3.
view has a higher load on rack2.

with RF + 1 == racks, base and view will have presence on all racks, so 3 pairs will be confined within a rack. in example above, only base:n31 and view:n22 will paired across racks.
with RF == racks, base and view will have one replica in each rack, so pairing is rack aware.
with RF - 1 == racks, at least one base and view replica will be paired within one rack.

could this optimistic approach work by just sorting base and view replicas within one DC by rack? the idea is to have rack aware pairing always enabled regardless of the RF and racks settings.

Reply:
it seems the patch is working on the assumption that the load over racks is always the same if RF % racks == 0, but I have a feeling it can be broken when adding / removing racks. so we'll end up with a crash, since base and view won't have the same number of replicas for a given rack.

Reply:
see that replicas cannot be naturally sorted by racks

```
cqlsh> select host_id,datacenter,rack from system.topology;

 host_id                              | datacenter | rack
--------------------------------------+------------+-------
 749f436e-c3d7-438a-b329-711f86dbf553 |       mydc | rack1
 88a311bc-651b-4f4c-a415-f34141b8fc82 |       mydc | rack2

 
cqlsh> CREATE KEYSPACE ks2 WITH replication = {'class': 'NetworkTopologyStrategy',                     'replication_factor': 1 } AND tablets = {'initial': 2 };
cqlsh> CREATE TABLE ks2.users (user_id bigint, country bigint, name text, PRIMARY KEY (user_id, country));

cqlsh> select table_id,replicas from system.tablets;

 table_id                             | replicas
--------------------------------------+----------------------------------------------------------------------------------------
 0a8285f0-9d2d-11ef-b4ff-6b2ba5fe103d | [(749f436e-c3d7-438a-b329-711f86dbf553, 0)]
 0a8285f0-9d2d-11ef-b4ff-6b2ba5fe103d | [(88a311bc-651b-4f4c-a415-f34141b8fc82, 0)]


cqlsh> alter keyspace ks2 WITH replication = {'class': 'NetworkTopologyStrategy',                     'mydc': 2 };
 
cqlsh> select table_id,replicas from system.tablets;

 table_id                             | replicas
--------------------------------------+----------------------------------------------------------------------------------------
 0a8285f0-9d2d-11ef-b4ff-6b2ba5fe103d | [(749f436e-c3d7-438a-b329-711f86dbf553, 0), (88a311bc-651b-4f4c-a415-f34141b8fc82, 1)]
 0a8285f0-9d2d-11ef-b4ff-6b2ba5fe103d | [(88a311bc-651b-4f4c-a415-f34141b8fc82, 0), (749f436e-c3d7-438a-b329-711f86dbf553, 1)]
```

Reply:
@raphaelsc can we resolve this thread?